src/runtimes/node.js

```diff
@@ -116,30 +117,47 @@ const zipFunction = async function ({
     ignoredModules: [...ignoredModulesFromConfig, ...ignoredModulesFromSpecialCases],
     srcFile: mainFile,
   })
+  const bundlerWarnings = data.warnings.length === 0 ? undefined : [data.warnings]
 
   // We're adding the bundled file to the zip, but we want it to have the same
-  // name and path as the original, unbundled file. For this, we use a rename.
-  const renames = {
-    [bundlePath]: mainFile,
+  // name and path as the original, unbundled file. For this, we use an alias..
+  const aliases = {
+    [mainFile]: bundlePath,
   }
   const basePath = commonPathPrefix([...dirnames, dirname(mainFile)])
 
   try {
     await zipNodeJs({
+      aliases,
       basePath,
       destFolder,
       destPath,
       filename,
       mainFile,
       pluginsModulesPath,
-      renames,
-      srcFiles: [...srcFiles, bundlePath],
+      srcFiles,
     })
   } finally {
     await cleanTempFiles()
   }
 
-  return destPath
+  return { bundlerVersion: JS_BUNDLER_ESBUILD, bundlerWarnings, path: destPath }
+}
+
+const zipWithFunctionWithFallback = async (parameters) => {
+  // If a specific JS bundler version is specified, we'll use it.
+  if (parameters.jsBundler) {
+    return zipFunction(parameters)
+  }
+
+  // Otherwise, we'll try to bundle with v2 and, if that fails, fallback to v1.
+  try {
+    return await zipFunction({ ...parameters, jsBundler: JS_BUNDLER_ESBUILD })
+  } catch (error) {
+    const data = await zipFunction({ ...parameters, jsBundler: JS_BUNDLER_LEGACY })
```

Comment:
If both `esbuild` and the legacy bundler fail, this probably mean there is an issue that is not related to the choice of bundler. In that case, would it make sense to re-throw the esbuild `error` (as opposed to the exception thrown by the legacy bundler)?

Reply:
I think that's a good idea. I've done it in https://github.com/netlify/zip-it-and-ship-it/pull/339/commits/ee896198d1a952e83283a8865aa770cf8e6ad8cb#diff-dda86982a0cad48036d5871c2d2f251ac96f1b30089e1e4879cd9903eec8cbcbR153-R164. Let me know if that's what you had in mind.

Reply:
Yes, exactly! :+1: